### PR TITLE
doc: add `__structuredAttrs`, `outputChecks`, `unsafeDiscardReferences`

### PR DESCRIPTION
**Pull request by ncfavier**
Follow-up to https://github.com/NixOS/nix/pull/7087, fixes https://github.com/NixOS/nix/issues/7429 (@roberth), fixes https://github.com/NixOS/nix/issues/7537.

screenshot

### CODE REVIEW

**Review by thufschmitt — APPROVED** [review]
Thanks a lot for taking the time to write this down!

File: doc/manual/src/language/advanced-attributes.md [hunk 1]
```diff
@@ -255,3 +255,67 @@ Derivations can declare some infrequently used optional attributes.
     > substituted. Thus it is usually a good idea to align `system` with
     > `builtins.currentSystem` when setting `allowSubstitutes` to
     > `false`. For most trivial derivations this should be the case.
+
+  - [`__structuredAttrs`]{#adv-attr-structuredAttrs}\
+    If the special attribute `__structuredAttrs` is set to `true`, the other derivation
+    attributes are serialised in JSON format and made available to the
+    builder via the file `.attrs.json` in the builder’s temporary
+    directory. This obviates the need for `passAsFile` since JSON files
```
Comment (fricklerhandwerk) [inline doc/manual/src/language/advanced-attributes.md:263]: Please add a link to `passAsFile`.

File: doc/manual/src/language/advanced-attributes.md [hunk 2]
```diff
@@ -255,3 +255,67 @@ Derivations can declare some infrequently used optional attributes.
...
+  - [`outputChecks`]{#adv-attr-outputChecks}\
+    When using [structured attributes](#adv-attr-structuredAttrs), the `outputChecks`
+    attribute allows defining checks per-output.
+
+    In addition to
+    [`allowedReferences`](#adv-attr-allowedReferences), [`allowedRequisites`](#adv-attr-allowedRequisites),
+    [`disallowedReferences`](#adv-attr-disallowedReferences) and [`disallowedRequisites`](#adv-attr-disallowedRequisites),
+    the following attributes are available:
+
+    - `maxSize` defines the maximum size of the output path.
```
Comment (fricklerhandwerk) [inline doc/manual/src/language/advanced-attributes.md:285]:
```suggestion
- `maxSize` defines the maximum size of the resulting [store object](../glossary.md#gloss-store-object).
```

Saying that a path has a size makes me think that we somehow mean the length of the string signifying the path, which doesn't make sense. It's the store object that can have a size.

File: doc/manual/src/language/advanced-attributes.md [hunk 3]
```diff
@@ -255,3 +255,67 @@ Derivations can declare some infrequently used optional attributes.
...
+    - `maxClosureSize` defines the maximum size of the output's closure.
+    - `ignoreSelfRefs` controls whether self-references should be considered when
+      checking for allowed references/requisites.
+
```
Comment (fricklerhandwerk) [inline doc/manual/src/language/advanced-attributes.md:289]:
```suggestion

Example:

```

File: doc/manual/src/language/advanced-attributes.md [hunk 4]
```diff
@@ -255,3 +255,67 @@ Derivations can declare some infrequently used optional attributes.
...
+    outputChecks.dev = {
+      # The 'dev' output must not be larger than 128 KiB.
+      maxSize = 128 * 1024;
+    };
+    ```
+
+  - [`unsafeDiscardReferences`]{#adv-attr-unsafeDiscardReferences}\
+    When using [structured attributes](#adv-attr-structuredAttrs), the **experimental**
+    attribute `unsafeDiscardReferences` is a per-output boolean which, if set to `true`,
+    disables scanning the build output for runtime dependencies altogether.
```
Comment (fricklerhandwerk) [inline doc/manual/src/language/advanced-attributes.md:310]:
```suggestion

    > **Warning**
    > This is an experimental feature.
    >
    > To enable it, add the following to [nix.conf](../command-ref/conf-file.md):
    >
    > ```
    > extra-experimental-features = discard-references
    > ```

    When using [structured attributes](#adv-attr-structuredAttrs), the
    attribute `unsafeDiscardReferences` is an attribute set with a boolean value for each output name.
    If set to `true`, it disables scanning the output for runtime dependencies.
    
    Example:
```

File: doc/manual/src/language/advanced-attributes.md [hunk 5]
```diff
@@ -255,3 +255,67 @@ Derivations can declare some infrequently used optional attributes.
...
+    ```nix
+    __structuredAttrs = true;
+    unsafeDiscardReferences.out = true;
+    ```
+
+    This is useful, for example, when generating self-contained filesystem images with
+    their own embedded Nix store: hashes found inside such an image refer
+    to the embedded store and not to the host's Nix store.
+
+    This is only allowed if the `discard-references` experimental feature is enabled.
```
Comment (fricklerhandwerk) [inline doc/manual/src/language/advanced-attributes.md:321]:
```suggestion
```

**Review by fricklerhandwerk — COMMENTED** [review]
@ncfavier sorry for the post-factum review, I was on vacation. Would be great if you could make a follow-up to address the suggestions you agree with. Those are for having a more predictable presentation of notes and examples, and, of course, adding links.

**Pull request merged**